### PR TITLE
Implement an ElasticTrainer to keep batch size fixed.

### PR DESCRIPTION
**Pull request by workingloong**

### CODE REVIEW

File: dlrover/trainer/torch/elastic.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    """
+    Singleton class that has information related to gradient
+    synchronization for gradient accumulation.
+
+    **Available attributes:**
+        - *sync_gradients** (`bool`) -- Whether gradients are currently
+            being synchronized across all processes.
+        - *num_backward_steps** (`int`) -- The number of step to perform
+            backward to compute gradients.
+        - *num_steps** (`int`) -- The number of step to update gradients.
```
Comment (skydoorkai) [inline dlrover/trainer/torch/elastic.py:36]: parameter update step

File: dlrover/trainer/torch/elastic.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+    def __init__(self):
+        self.__dict__ = self._shared_state
+        if not self.initialized:
+            self.sync_gradients = True
+            self.num_backward_steps = 0
+            self.num_steps = 0
+
+    def check_sync_gradient(self, gradient_accumulation_steps):
+        """Check wether to synchronize gradients accross all processes."""
```
Comment (skydoorkai) [inline dlrover/trainer/torch/elastic.py:49]: whether

File: dlrover/trainer/torch/elastic.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    def __getstate__(self):
+        return self.__dict__.copy()
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+
+class _ElasticLRScheduler(object):
+    """A wrapper around a learning rate scheduler that will only
+    step after the gradients are sychronizec across all processes and
```
Comment (skydoorkai) [inline dlrover/trainer/torch/elastic.py:124]: sychronizing

File: dlrover/trainer/torch/elastic.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        """The job master of DLRover will notify the worker to do chekpoint
+        before it changes workers."""
+        pass
+
+    def _set_gradient_accumulation_steps(self):
+        max_worker_num = int(os.getenv("WORKER_NUM", 0))
+        if max_worker_num == 0:
+            self.gradient_accumulation_steps = 1
+
+        cur_world_size = int(os.getenv("WORLD_SIZE", 1))
```
Comment (skydoorkai) [inline dlrover/trainer/torch/elastic.py:261]: Who set WORLD_SIZE/RANK env?

Maybe better to use dist.get_world_size/get_rank.

Also, this assumes that data parallel is used. If other parallel method is used, this logic is incorrect.

**Comment by hxdtest** [comment]
LGTM
Add a doc and examples（killing a pod during training）later.

**Review by hxdtest — APPROVED** [review]

**Pull request merged**